mitodl/mit-open
# daily subscription email to subscribers

### What are the relevant tickets?

Closes #848

### Description (What does it do?)
This PR adds functionality that sends a daily digest email to a user for new learning that match their subscriptions

### Screenshots
<img width="920" alt="Screenshot 2024-05-20 at 4 14 59 PM" src="https://github.com/mitodl/mit-open/assets/196425/e6185a8e-d5a0-4ef0-80da-d5548976423a">


### How can this be tested?
1. Checkout this branch
2. populate your local with courses and topic channels
```
python manage.py update_topics
python manage.py update_offered_by
python manage.py backpopulate_mitxonline_data
python manage.py backpopulate_micromasters_data
python manage.py backpopulate_resource_channels
python manage.py backpopulate_xpro_data
python manage.py backpopulate_ocw_data --skip-contentfile
python manage.py recreate_index --all
```
3. Head to a channel page and subscribe to it (http://localhost:8063/c/department/physics)
4. pick one or two resources from that topic page and note the id(s) - it would be a good test to subscribe to several channels and have multiple resources that you note
5. in django shell, manually set the "created_on" attribute to today so that the celery task picks them up as 'newly created' resources - replace the id in the following with the resources noted in the last step
```
from learning_resources.models import LearningResource
from main.utils import now_in_utc
lr = LearningResource.objects.get(id=438)
lr.created_on = now_in_utc()
lr.save()
```
6. re-run ```python manage.py recreate_index --all``` to force sync the index
7. in django shell, import the celery task and manually run it via:
```
from learning_resources_search.tasks import send_subscription_emails
send_subscription_emails('channel_subscription_type')
```
You should see the contents of the email outputted in the docker logs of the web container. The resources should be grouped by the channels.

### Additional Context
to visually see the emails,  I have been modifying the send_temlate_email method in profiles/utils.py like so
```
 def send_template_email(recipients, subject, template, context):
+    from main.utils import chunks, merge_strings, now_in_utc
     """
     Send an html email using a provided template
     Args:
     if not context:
         context = {}
     html_content = render_to_string(template, context)
+    with open(f"{now_in_utc()}.html", "w") as f:
+        f.write(html_content)
```
This will output an html file for each email being sent which you can view in a browser

### Checklist:
- [ ] subscription emails have been tested with multiple resources and users
